Rustls client auth

yedayak opened the pull request.

icing committed:
conncache: count shutdowns against host and max limits

Count connections to a host against a possibly configured destination
limit. Trigger multi `connchange` when a connection has been shutdown,
so pending transfers can try to get a connection once again.

Reported-by: baranyaib90 on github
Fixes #15857
Closes #15879

yedayak committed:
rustls: use client cert and key if available

nhorman committed:
osslq: use SSL_poll to determine writeability of QUIC streams

This discussion:
https://github.com/openssl/openssl/discussions/23339#discussion-6094341

Specifically item number 2 (Send Blocking) was raised by the curl team,
noting that SSL_want_write returning false was not a good indicator of
when a stream is writeable. The suggestion in that discussion was to use
SSL_poll with an SSL_POLL_EVENT_W flag instead, as that is a proper
indication of when an SSL_object will allow writing without blocking.

While ssl_want_write updates its state based on the last error
encountered (implying a need to retry an operation to update the
last_error state again), SSL_poll checks stream buffer status during the
call, giving it more up to date information on request. This is the
method used by our guide demos (quic-hq-interop specifically), and it
works well.

This change has been run through the curl test suite, and shown to pass
all tests. However, given the initial problem description I'm not sure
if there is a test case that explicitly checks for blocking and
unblocking of streams. As such some additional testing may be warranted.

Closes #15909

monnerat committed:
mime: explicitly rewind subparts at attachment time.

Subparts may have been previously used as a top-level mime structure and
thus not rewound.

New test 695 checks the proper functioning in these particular conditions.

Reported-by: Qriist on github
Fixes #15842
Closes #15911

yedayak committed:
rustls: client auth: propogate rustls errors

yedayak committed:
rustls: Add fixme regarding keys_match

yedayak committed:
docs/examples/checksrc: don't allow snprintf specifically

This isn't needed anymore after https://github.com/curl/curl/pull/15835,
since banned functions are just allowed in general in
`docs/examples/.checksrc`, and emits a warning when running make
checksrc:

`invalid warning specified in .checksrc: "SNPRINTF"`

Closes #15916

bagder committed:
RELEASE-NOTES: synced

dependabot[bot] committed:
GHA: bump cygwin/cygwin-install-action from 4 to 5

Bumps [cygwin/cygwin-install-action](https://github.com/cygwin/cygwin-install-action) from 4 to 5.
- [Release notes](https://github.com/cygwin/cygwin-install-action/releases)
- [Commits](https://github.com/cygwin/cygwin-install-action/compare/006ad0b0946ca6d0a3ea2d4437677fa767392401...f61179d72284ceddc397ed07ddb444d82bf9e559)

---
updated-dependencies:
- dependency-name: cygwin/cygwin-install-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Closes #15918

renovate[bot] committed:
GHA/linux: update awslabs/aws-lc to v1.42.0

Closes #15919

vszakats committed:
cmake: deprecate winbuild, add migration guide from legacy build methods

We recommend migrating to CMake from winbuild and Visual Studio project
files. winbuild is deprecated and will be dropped in September 2025.

CMake supports all the features and options, with new ones added
promptly. It supports out-of-tree, unity and documentation builds.

- deprecate winbuild method in favour of CMake by September 2025.
- add migration guide from winbuild to CMake.
- add migration guide from Visual Studio Project Files to CMake.
- add deprecation message to winbuild.
  Need to ack with `WINBUILD_ACKNOWLEDGE_DEPRECATED=yes`
  Authored-by: Jay Satiro
- mention `CMAKE_BUILD_TYPE` option in `INSTALL-CMAKE`.
- document missing `SSH_PATH` winbuild option.

Closes #15920

bagder committed:
cookie: small variable type cleanups

- bump cookie counter and "creation time" to use 'unsigned int'
- use BIT() for single-bit struct field
- make invalid_octets() return bool properly

Closes #15921

vszakats committed:
msvc: fix building with `HAVE_INET_NTOP` and MSVC <=1900

MSVC 1900 and older is missing a `const` specifier in the `inet_ntop()`
declaration for the second argument. A workaround was in place for it
in cmake, but it didn't cover all necessary versions.

Replace the workaround with a different one, move it to `lib/inet_ntop.c`
and extend to all necessary MSVC versions.

Also add CI jobs for the older MSVC versions: 2013, 2015, 2017.

Closes #15923

vszakats committed:
cmake: pre-fill `HAVE_STDINT_H` on Windows

Closes #15925

icing committed:
vtls: feature ssls-export for SSL session im-/export

Adds the experimental feature `ssls-export` to libcurl and curl for
importing and exporting SSL sessions from/to a file.

* add functions to libcurl API
* add command line option `--ssl-sessions <filename>` to curl
* add documenation
* add support in configure
* add support in cmake
+ add pytest case

Closes #15924

vszakats committed:
msvc: add VS2012 CI job, fix compiler warnings in LDAP and tests

Follow-up to 8cfc936f5c9d1fe9a99956a916f95e137476b27c #10674
Follow-up to d620ec67913b2c4be25e7d72f1933f0a8a3e0a2e #14700

Closes #15926

vszakats committed:
appveyor: fixup job name

Follow-up to 50f6a6b1d419c9e6ebc4fb2a848fa20def622d38 #15926

vszakats committed:
GHA/windows: drop MSH3 job (broken after 0.7.0 bump)

Starting GHA runner image 20250105.1.0.

As seen on Linux with 0.7.0:
```
/home/runner/msh3/include/msh3.h:377:18: error: width of ‘RESERVED’ exceeds its type
  377 |             bool RESERVED                 : 5;
      |                  ^~~~~~~~
/home/runner/msh3/include/msh3.h:490:18: error: width of ‘RESERVED’ exceeds its type
  490 |             bool RESERVED            : 7;
      |                  ^~~~~~~~
```
https://github.com/curl/curl/actions/runs/12655717818/job/35266716846#step:35:195

Bug: https://github.com/curl/curl/pull/15924#issuecomment-2575106711
Bug: https://github.com/curl/curl/pull/15930#issuecomment-2575842486

Closes #15927

tyilo committed:
completion.pl: add completion for paths after @ for fish

Closes #15928

vszakats committed:
content_encoding: namespace GZIP flag constants

To avoid collision with a 3rd-party `RESERVED` symbols.

This symbol is used in the public header of MSH3 0.7.0.

Closes #15929

bagder committed:
DEPRECATE: remove msh3 in six months

The msh3 backed for QUIC and HTTP/3 was introduced in April 2022 but has
never been made to work properly. It has seen no visible traction or
developer activity from the msh3 main author (or anyone else seemingly
interested) in two years. As a non-functional backend, it only adds
friction and "weight" to the development and maintenance.

Meanwhile, we have a fully working backend in the ngtcp2 one and we have
two fully working backends in OpenSSL-QUIC and quiche well on their way
of ending their experimental status in a future.

We remove msh3 support from the curl source tree in July 2025.

Closes #15931

vszakats committed:
appveyor: reorder builds to get useful results earlier

Also to align with existing VS2010. Keeping the VS2008 job first to give
a quick sniff test for MSVC builds.

Follow-up to 08ff33e483e15b003de169cef33ca7cb6b6a32c0 #15923
Follow-up to 50f6a6b1d419c9e6ebc4fb2a848fa20def622d38 #15926

Closes #15932

renovate[bot] committed:
ci: update rojopolis/spellcheck-github-actions digest to 9e0a5fb

Closes #15933

vszakats committed:
appveyor: bump VS2008 jobs to VS2010

VS2008 has been partly broken for a while with its shared-debug builds
crashing on startup. Its compiler output (UTF-16 HTML) was also barely
readable even after conversion. It's also the only platform in CI
missing `stdint.h`.

This patch migrates a VS2008 job to VS2010 and drops another that
already had a VS2010 equivalent.

We recommend switching to VS2010 or newer when using MSVC to build curl.

Ref: #15907
Closes #15934

vszakats committed:
appveyor: always use cmake `-A` option to select x64

The `Win64` generator suffix alternative was required by old CMake
versions (<3.1) only:
https://cmake.org/cmake/help/v3.22/generator/Visual%20Studio%2010%202010.html

Closes #15935

bagder committed:
cookie: cap expire times to 400 days

The pending cookie RFC update (currently known as 6265bis draft-19) says

  Let cookie-age-limit be the maximum age of the cookie (which name of
  Max-Age and an attribute-value of expiry-time. SHOULD be 400 days or
  less.

This change makes received cookies over the wire get capped to 400 days.

It does not cap the expiry date of cookies loaded from file.

It does this by rounding the expire time to a even minute. This, to
allow the test suite to do the same and have a chance to get the same
number for stable testing without requiring a debug build.

The test script generates TWO numbers in the output file for each
%days[] used in the input test file, and the function that subsequently
compares and verifies output is fine with *either* of the two numbers.

This is done so that if the test case is generated the second
immediately before curl runs, that updated expiry number is also deemed
okay. It still checks for an exact match of either number.

Closes #15937

vszakats committed:
build: delete `-Wsign-conversion` related FIXMEs

We decided last year not to pursue avoiding this warning, because it
adds noise and friction, while in most cases not revealing actual code
issues. We fixed the interesting portion of them throughout mid-2024.

Conclude this effort by deleting related FIXMEs and temporary comments.

Follow-up to 3829759bd042c03225ae862062560f568ba1a231 #12489
Closes #15939

vszakats committed:
appveyor: tidy up shell code

Closes #15940

icing committed:
HTTP/2: strip TE request header

The TE request header field is invalid in HTTP/2. Since clients may not
know in advance if a connection negotiates HTTP/2, automatically strip
such a header when h2 is in play.

Add test_01_10 to verify.

Reported-by: Jiri Stary
Fixes #15941
Closes #15943

vszakats committed:
tidy-up: delete stray references from autotools, docs

- autotools: delete stray `VC14_LIB*` references.
- autotools: delete (now) empty `CLEANFILES`.
- autotools: delete no longer used lib/src .inc includes in root makefile.
- autotools: delete stray `cygwinbin` target.
- autotools: delete stray `pkgadd` target (Solaris).
- lib, src: delete stray files from `.gitignore`.
- INSTALL.md: delete reference to non-existing `src/config-win32.h`.
- lib/config-win32ce.h: whitespace.
- lib/config-win32ce.h: sync comments with `config-win32.h`.

Closes #15944

sftcd committed:
ECH: update APIs to those agreed with OpenSSL maintainers

Closes #15945

bagder committed:
scripts/delta: add number of days since first httpget release

vszakats committed:
msvc: drop checks for ancient versions

- drop version guard for `__inline`.
  Supported since `_MSC_VER` 1000.
  Visual C++, 32-bit, version 4.0 (1996)

- drop version guard for `__declspec(noreturn)` and `__forceinline`.
  Supported since `_MSC_VER` 1200.
  Visual C++, 32-bit, version 6.0 (1998)

For ancient versions, it's possible to override the default behaviour
by setting these macros via `CPPFLAGS`: `CURL_NORETURN`, `CURL_INLINE`,
`CURL_FORCEINLINE`

Closes #15946

vszakats committed:
cmake: update a comment [ci skip]

Follow-up to 421e592db25cbbe4baadfeef8e6ed75a57579d0a #15832

talregev committed:
GHA/windows: add wolfSSL to MSVC MultiSSL job

Follow-up to 98932f34879bba86339b8ca94ba04aa994c744f8 #15765

Closes #15438

bagder committed:
vtls_scache: avoid a "Redundant Condition"

Pointed out by CodeSonar. "ssl_config" can in fact not be NULL here.
Made it an assert instead.

Closes #15948

bagder committed:
tool_ssls: fix "ignored return value" warning

Pointed out by CodeSonar. While harmless, we might as well address it.

Closes #15949

bagder committed:
vtls_scache: fix possible null ptr deref

Pointed out by CodeSonar. It is probably not reachable, but might as
well just add a precaution.

Closes #15950

bagder committed:
curl_trc: repair build with verbose strings disabled

- moved the dummy functions into the C file, made them non-static
- added a Curl_trc_ssls dummy

Closes #15951

icing committed:
ssl session cache: change cache dimensions

Use a larger one when shared.

Closes #15953

vszakats committed:
msvc: drop checks for `_MSC_VER >= 900`

It's Visual C++, 32-bit, version 2.0, released in 1993. Used to verify
if `_INTEGRAL_MAX_BITS` is available.

After this patch we assume `_INTEGRAL_MAX_BITS` is always available in
MSVC.

Closes #15952

icing committed:
GnuTLS: fix 'time_appconnect' for early data

When using early data with GnuTLS, the the timer `appconnect`
had the value from the "pretended" connect, not when the actual
TLS handshake was done.

Closes #15954

vszakats committed:
msvc: assume `_INTEGRAL_MAX_BITS >= 64`

curl requires C compilers supporting 64-bit `curl_off_t` type since
835682661c7a36b0e860c978204e8cd9b033be45 #10597 (v8.0.0).

Assume the MSVC compiler offers the necessary support.

It makes curl require Visual Studio .NET 2003, v7.1 (`_MSC_VER = 1310`).
With the possibility that 1300 (Visual Studio .NET, v7.0, 2002), or 1200
(Visual C++, 32-bit, v6.0, 1998) may also work.

Follow-up to ca18198dd4d6517bd2d8b34a99fcd9d4f780d219 #15952
Closes #15955

vszakats committed:
system.h: drop compilers lacking 64-bit integer type (Windows/MS-DOS)

- DJGPP 1.x (including `__GO32__`) (MS-DOS)
  DJGPP 2.x support remains unchanged.
- Salford C (Windows)
- Turbo C (Windows 16-bit)
- Borland C++ < 5.2 (Windows 16-bit?)
- Pelles C < 2.8 (Windows)

These targets mapped `curl_off_t` to `long`. On Windows and MS-DOS
`long` is always 32-bit.

curl requires C compilers supporting 64-bit `curl_off_t` type since
835682661c7a36b0e860c978204e8cd9b033be45 #10597 (v8.0.0).

Also: drop remaining `__GO32__` and Salford C guards.

Closes #15957

yedayak commented:
Hi @cpu, hope you don't mind the ping.
Do you think this makes sense open as a PR for curl even before `rustls_certified_key_keys_match` exists? Or should this wait as well until the next rustls-ffi release?

bagder committed:
RELEASE-NOTES: synced

bagder committed:
RELEASE-NOTES: bump cmdline options and number of public funcs

vszakats committed:
build: fix `-Wtrampolines` picky warning for gcc 4.x versions

Caused an error when using cmake with gcc 4.4 and 4.5.
Also 4.3 when using autotools.

Seen with GNU 4.4.0 in CeGCC 0.59.1:
```
cc1: error: unrecognized command line option "-Wtrampolines"
```

`-Wtrampolines` requires gcc 4.6 or upper.

Ref: https://gcc.gnu.org/onlinedocs/gcc-4.6.0/gcc/Warning-Options.html
Ref: https://master.dl.sourceforge.net/project/cegcc/cegcc/0.59.1/cegcc_mingw32ce_snowleopard_r1397.tar.bz2

Closes #15962

cpu commented:
> Hi @cpu, hope you don't mind the ping.

Happy to be pinged :-) Thanks for taking a look at this work.

> Do you think this makes sense open as a PR for curl even before rustls_certified_key_keys_match exists? Or should this wait as well until the next rustls-ffi release?

My personal opinion is that it makes sense to stage this work but not open it as a PR for the Curl developers until there is a rustls-ffi release that offers the needed functionality.

I've been tracking the last bits needed for a release in https://github.com/rustls/rustls-ffi/pull/518 - I'm mostly waiting on some upstream Rustls work for post-qc kx.

Does that make sense to you? I hope it will only be a matter of weeks and not months.

renovate[bot] committed:
ci: update actions/upload-artifact digest to 65c4c4a

Closes #15964

vszakats committed:
system.h: drop duplicate and no-op code

- drop duplicate declaration block under `__MVS__`.
- drop duplicate declaration block under `__IBMC__`.
- drop empty `#if` branches under `__IBMC__`.

Closes #15966

kpcyrd commented:
FWIW, rustls-ffi 0.15.0 has now been released. :)

cpu commented:
In case it wasn't obvious, I rolled the content of this PR into https://github.com/curl/curl/pull/16828 (but forgot to link here. Fixed that!): https://github.com/curl/curl/pull/16828/commits/14a89d28a9490fd62e7482f947b94a3ccb2b51da

cpu commented:
This work was merged into curl w/ https://github.com/curl/curl/commit/1c8c93ae15c692c547e3238c4f067f76616a53e8

Thanks again :-)
